### ISSUE TITLE
feat: allow different ways of querying code content

### ISSUE THREAD

**Issue by jmbuhr** [jmbuhr/otter.nvim#37]
- Need to decouple `@lang` from `@code` to allow for language-specific queries, where the type of query already defines `lang` and doesn't return it https://github.com/jmbuhr/otter.nvim/blob/5bfc1f5ab9cd19647dc7a919694fab7c1cfe0ccb/lua/otter/tools/queries.lua
- Extend the API even further to allow for custom ways of getting code ranges

**Comment by wouthoekstra** [comment]
Treesitter supports syntax highlighting for nested languages by adding a query to your nvim configuration, by putting it in a queries/LANGUAGE/injections.scm file. Would it be possible to use these same queries?

For example; I made a query to get syntax highlighting for SQL in my Javascript files by creating the following query in ~/.config/nvim/queries/javascript/injections.scm. Because the output variable is '@sql', treesitter recognizes it as sequel and highlights it as such:

```
(call_expression
  (member_expression
    object: (identifier) @_object (#eq? @_object "knex")
    property: (property_identifier) @_property (#eq? @_property "raw")
  )

  (arguments
    (template_string) @sql
    (#offset! @sql 0 1 0 -1)
  )
)
```

**Comment by jmbuhr** [comment]
Indeed, my dream-API would be to get the ranges of injections directly from the treesitter module in neovim. Upon closer searching, I think this is possible.
https://github.com/neovim/neovim/blob/58618d208acd3827c4e86668529edb619bb9b8dd/runtime/lua/vim/treesitter/query.lua#L210 can get the query for all injections and I believe this includes user-defined ones.

And from this we should be able to get those captures:
https://github.com/neovim/neovim/blob/58618d208acd3827c4e86668529edb619bb9b8dd/runtime/doc/treesitter.txt#L484-L505

**Comment by wouthoekstra** [comment]
Wow this is great! For me it still does not properly work however, and that might be out of the scope of your package. But the used query injection does not seem to acknowledge the offset I've defined for querying my sql. Which means the backticks in my code are incorrectly parsed as part of the SQL:

For example, the query in my previous comment finds the SQL inside the knex.raw(), but it trips over the backticks:
    const plan = await knex.raw(`
      SELECT plan.name
        FROM subscription
        JOIN plan ON subscription.plan_id = plan.id
       WHERE account_id = :accountId
   `, { accountId: user.accountId }).then(firstOrNull);

This might be more an issue with treesitter, but I am not sure.

**Comment by jmbuhr** [comment]
Looks like the `offset` directive only affects the `range` reported by the nodes metadata, not the text returned by `get_node_text` as currently used (https://github.com/jmbuhr/otter.nvim/blob/242d180e7f23cc2af6b5d5193bc50909408caef7/lua/otter/keeper.lua#L67).
However, I can't get treesitter to give me the correct range with directives applied, which supposedly was supposed to happen with the way I do it in the commit mentioned above. Any idea?

**Comment by wouthoekstra** [comment]
If I read this correctly, someone proposed a change to treesitter to update the injection with the offset: https://github.com/nvim-treesitter/nvim-treesitter/pull/1240/files#diff-81f58ecd0f4978b575136c86619adc8dbc9a0c9e87f87702ddf22615d0ac6c4bR93-R102. But it is already two years old.

**Comment by jmbuhr** [comment]
I might be able to revisit this with the upstream changes to injections
